tf_encrypted/keras/layers/dense.py

```diff
@@ -0,0 +1,121 @@
+      units: Positive integer, dimensionality of the output space.
+      activation: Activation function to use.
+          If you don't specify anything, no activation is applied
+          (ie. "linear" activation: `a(x) = x`).
+      use_bias: Boolean, whether the layer uses a bias vector.
+      kernel_initializer: Initializer for the `kernel` weights matrix.
+      bias_initializer: Initializer for the bias vector.
+      kernel_regularizer: Regularizer function applied to
+          the `kernel` weights matrix.
+      bias_regularizer: Regularizer function applied to the bias vector.
+      activity_regularizer: Regularizer function applied to
+          the output of the layer (its "activation")..
+      kernel_constraint: Constraint function applied to
+          the `kernel` weights matrix.
+      bias_constraint: Constraint function applied to the bias vector.
+
+  Input shape:
+      nD tensor with shape: `(batch_size, ..., input_dim)`.
+      The most common situation would be
+      a 2D input with shape `(batch_size, input_dim)`.
+
+  Output shape:
+      nD tensor with shape: `(batch_size, ..., units)`.
+      For instance, for a 2D input with shape `(batch_size, input_dim)`,
+      the output would have shape `(batch_size, units)`.
+  """
+
+  def __init__(self,
+               units,
+               activation=None,
+               use_bias=True,
+               kernel_initializer='glorot_uniform',
+               bias_initializer='zeros',
+               kernel_regularizer=None,
+               bias_regularizer=None,
+               activity_regularizer=None,
+               kernel_constraint=None,
+               bias_constraint=None,
+               **kwargs):
+
+    super(Dense, self).__init__(**kwargs)
+
+    self.units = int(units)
+    self.activation = activations.get(activation)
+    self.use_bias = use_bias
+
+    self.kernel_initializer = initializers.get(kernel_initializer)
+    self.bias_initializer = initializers.get(bias_initializer)
+
+    if kernel_regularizer:
+      raise NotImplementedError(argument_msg.format("kernel_regularizer",
+                                                    "Dense"))
+    if bias_regularizer:
+      raise NotImplementedError(argument_msg.format("bias_regularizer",
+                                                    "Dense"))
+    if activity_regularizer:
+      raise NotImplementedError(argument_msg.format("activity_regularizer",
+                                                    "Dense"))
+    if kernel_constraint:
+      raise NotImplementedError(argument_msg.format("kernel_constraint",
+                                                    "Dense"))
+    if bias_constraint:
+      raise NotImplementedError(argument_msg.format(" bias_constraint",
+                                                    "Dense"))
+
+  def compute_output_shape(self, input_shape):
+    return [input_shape[0] + self.units]
+
+  def build(self, input_shape) -> None:
+    units_in = int(input_shape[1])
```

Comment:
same question as in `compute_output_shape`

Reply:
this is going to break if the input_shape has rank greater than 2

Reply:
I'd also recommend an assert here that checks to make sure units is equal to dimension 1 of the flattened input tensor

Reply:
and this means you can store the flattened input_shape as an attribute for use in `call`, so that can be simplified a bit

Reply:
ah good catch